casper/src/main/scala/coop/rchain/casper/MultiParentCasperImpl.scala

```diff
@@ -554,21 +554,75 @@ class MultiParentCasperImpl[F[_]: Sync: Capture: ConnectionsCell: TransportLayer
       dependencyFreeBlocks = blockBuffer
         .filter(block => dependencyFree.contains(block.blockHash))
         .toList
-      attempts <- dependencyFreeBlocks.traverse(b => attemptAdd(b))
+      attempts <- dependencyFreeBlocks.traverse { b =>
+                   for {
+                     status <- attemptAdd(b)
+                   } yield (b, status)
+                 }
       _ <- if (attempts.isEmpty) {
             ().pure[F]
           } else {
-            Capture[F].capture {
-              dependencyFreeBlocks.map {
-                blockBuffer -= _
-              }
-            } *>
-              blockBufferDependencyDagState.set(dependencyFree.foldLeft(blockBufferDependencyDag) {
-                case (acc, hash) =>
-                  DoublyLinkedDagOperations.remove(acc, hash)
-              }) *> reAttemptBuffer
+            for {
+              _ <- removeAdded(blockBufferDependencyDag, attempts)
+              _ <- reAttemptBuffer
+            } yield ()
           }
     } yield ()
 
+  private def removeAdded(
+      blockBufferDependencyDag: DoublyLinkedDag[BlockHash],
+      attempts: List[(BlockMessage, BlockStatus)]
+  ): F[Unit] =
+    for {
+      successfulAdds <- attempts
+                         .filter {
+                           case (_, status) => successfulAddStatus(status)
+                         }
+                         .pure[F]
+      _ <- unsafeRemoveFromBlockBuffer(successfulAdds)
+      _ <- removeFromBlockBufferDependencyDag(blockBufferDependencyDag, successfulAdds)
+    } yield ()
+
+  private def unsafeRemoveFromBlockBuffer(
+      successfulAdds: List[(BlockMessage, BlockStatus)]
+  ): F[Unit] =
+    Sync[F].delay {
+      successfulAdds.map {
+        blockBuffer -= _._1
+      }
+      ()
+    }
+
+  private def removeFromBlockBufferDependencyDag(
+      blockBufferDependencyDag: DoublyLinkedDag[BlockHash],
+      successfulAdds: List[(BlockMessage, BlockStatus)]
+  ): F[Unit] =
+    blockBufferDependencyDagState.set(
+      successfulAdds.foldLeft(blockBufferDependencyDag) {
+        case (acc, successfulAdd) =>
+          DoublyLinkedDagOperations.remove(acc, successfulAdd._1.blockHash)
+      }
+    )
+
+  private def successfulAddStatus(status: BlockStatus): Boolean =
+    status == Valid ||
+      status == AdmissibleEquivocation ||
+      status == IgnorableEquivocation ||
+      status == InvalidUnslashableBlock ||
+      status == MissingBlocks ||
+      status == InvalidBlockNumber ||
+      status == InvalidRepeatDeploy ||
+      status == InvalidParents ||
+      status == InvalidFollows ||
+      status == InvalidSequenceNumber ||
+      status == InvalidShardId ||
+      status == JustificationRegression ||
+      status == NeglectedInvalidBlock ||
+      status == NeglectedEquivocation ||
+      status == InvalidTransaction ||
+      status == InvalidBondsCache ||
+      status == InvalidBlockHash ||
+      status == InvalidDeployCount
```

Comment:
So everything except `Processing` and `BlockException` is a _successful add status_.  This is shorter:
```scala
status match {
  case Processing | BlockException => false
  case _ => true
}
```

Reply:
OK, only now I read:
>  So while it is a bit ugly, we whitelist instead of blacklist which blocks get successfully added during a reAttempt call.

This is a good call, because we are we will be obviously broken, rather than un-obviously insecure.

So how about adding a field to `BlockStatus`:
```scala
sealed trait BlockStatus {
  val addedSuccessfuly: Boolean
}

case object Processing                   extends BlockStatus {
  override  val addedSuccessfuly: Boolean = false
}
case class BlockException(ex: Throwable) extends BlockStatus {
  override  val addedSuccessfuly: Boolean = false
}

sealed trait ValidBlock   extends BlockStatus {
  override  val addedSuccessfuly: Boolean = true
}
sealed trait InvalidBlock extends BlockStatus {
  override  val addedSuccessfuly: Boolean = true
}
```
and now the this method becomes:
```scala
  private def successfulAddStatus(status: BlockStatus): Boolean = status.addedSuccessfuly
```

Reply:
Sounds good